linting/src/no_main.rs

```diff
@@ -0,0 +1,98 @@
+// Copyright (C) Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use ast::{
+    AttrStyle,
+    Crate,
+};
+use clippy_utils::diagnostics::span_lint_and_help;
+use if_chain::if_chain;
+use rustc_ast as ast;
+use rustc_lint::{
+    EarlyContext,
+    EarlyLintPass,
+    LintContext,
+};
+use rustc_middle::lint::in_external_macro;
+use rustc_session::{
+    declare_lint,
+    declare_lint_pass,
+};
+use rustc_span::sym;
+
+declare_lint! {
+    /// ## What it does
+    /// Checks if a contract is annotated with the `no_main` inner attribute.
+    ///
+    /// ## Why is this necessary?
+    /// Contracts must be annotated with `no_main` inner attribute when compiled for on-chain
+    /// execution.
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// // Bad: Contract does not contain the `no_main` attribute, so it cannot be compiled to Wasm
+    /// #![cfg_attr(not(feature = "std"), no_std)]
+    ///	#[ink::contract]
+    ///	mod my_contract { /* ... */ }
+    /// ```
+    ///
+    /// Use instead:
+    ///
+    /// ```rust
+    /// #![cfg_attr(not(feature = "std"), no_std, no_main)]
+    ///	#[ink::contract]
+    ///	mod my_contract { /* ... */ }
+    /// ```
+    pub NO_MAIN,
+    Deny,
+    "contract must be annotated with the `no_main` inner attribute"
+}
+
+declare_lint_pass!(NoMain => [NO_MAIN]);
+
+/// Returns true if the target architecture is suitable to be executed on-chain
+fn is_contract_build(cx: &EarlyContext<'_>) -> bool {
+    matches!(
+        cx.sess().target.llvm_target.to_string().as_str(),
+        "wasm32-unknown-unknown" | "riscv32i-unknown-none-elf"
+    )
+}
+
+impl EarlyLintPass for NoMain {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, krate: &Crate) {
+        // Disable when building for e2e tests
+        if !is_contract_build(cx) {
```

Comment:
Can we execute this lint on all the builds, even for e2e tests, making `no_main` mandatory for every ink! contract?

The point is that `cargo-contract` always runs linter for the host architecture. It is possible to hack it a bit, but should we? Why should we skip this lint for tests?

Reply:
E2E tests will still run the linter because they compile the contract code blob.

Maybe what we really mean/want is that when building the contract "natively" i.e. not `no_std`, which happens for unit tests and metadata generation. But those are usually executed on a different path so maybe we don't need to make the distinction here.

Reply:
Yes, I see. Let's try to introduce this lint for all contracts. I think it won't make any problems.